worker/codalabworker/local_run/local_run_state.py

```diff
@@ -0,0 +1,277 @@
+
+            dependency_path = self._run_manager.dependency_manager.get(
+                        bundle_uuid,
+                        (dep['parent_uuid'], dep['parent_path'])).path
+            dependency_path = os.path.join(self._run_manager.dependencies_dir, dependency_path)
+
+            docker_dependency_path = os.path.join(docker_dependencies_path, dep['child_path'])
+
+            os.symlink(docker_dependency_path, child_path)
+            # These are turned into docker volume bindings like:
+            #   dependency_path:docker_dependency_path:ro
+            dependencies.append((dependency_path, docker_dependency_path))
+
+        # 3) Set up container
+        if run_state.resources['request_network']:
+            docker_network = self._run_manager.docker_network_external_name
+        else:
+            docker_network = self._run_manager.docker_network_internal_name
+
+        try:
+            cpuset, gpuset = self._run_manager.assign_cpu_and_gpu_sets(
+                run_state.resources['request_cpus'], run_state.resources['request_gpus'])
+        except:
+            run_state.info['failure_message'] = "Cannot assign enough resources"
+            return run_state._replace(stage=LocalRunStage.CLEANING, info=run_state.info)
+
+        # 4) Start container
+        container_id = self._run_manager.docker.start_container(
+            run_state.bundle_path, bundle_uuid, run_state.bundle['command'],
+            run_state.resources['docker_image'], docker_network, dependencies,
+            cpuset, gpuset, run_state.resources['request_memory']
+        )
+
+        digest = self._run_manager.docker.get_image_repo_digest(run_state.resources['docker_image'])
+
+        return run_state._replace(stage=LocalRunStage.RUNNING, run_status='Running',
+            container_id=container_id, docker_image=digest, has_contents=True, cpuset=cpuset, gpuset=gpuset)
+
+    def _transition_from_RUNNING(self, run_state):
+        def check_and_report_finished(run_state):
+            try:
+                finished, exitcode, failure_msg = self._run_manager.docker.check_finished(run_state.container_id)
+            except DockerException:
+                traceback.print_exc()
+                finished, exitcode, failure_msg = False, None, None
+            return dict(finished=finished, exitcode=exitcode, failure_message=failure_msg)
+
+        bundle_uuid = run_state.bundle['uuid']
+
+        new_info = check_and_report_finished(run_state)
+        run_state.info.update(new_info)
+        run_state = run_state._replace(info=run_state.info)
+
+        if run_state.is_killed and run_state.container_id is not None:
+            try:
+                self._run_manager.docker.kill_container(run_state.container_id)
+            except DockerException:
+                traceback.print_exc()
+            return run_state._replace(stage=LocalRunStage.CLEANING, container_id=None)
```

Comment:
`is_killed` really means kill request is there, not that it is killed? maybe variable name should reflect this better

Should we transition into CLEANING?  What happens if the kill request failed?

Reply:
`is_killed` in this context is taken to mean the run was killed by the user server-side. So in a sense the run is already killed and now the run manager has to deal with it (this is the semantics we want to enforce) 

I don't know under what condition the kill_container call might fail and so I don't know how to handle that case. For all it's worth all this exception capturing was already there in the first place and I was reluctant to remove it, maybe we should? Transitioning to cleaning up is still the best option I think.